### PR TITLE
ci: Add check-generated-protobuf

### PR DESCRIPTION
**Pull request by dnephin**
This PR adds a CI job which checks that the generated Go files match the `.proto` files. 

With this change we ensure that:

1. all code is generated with the same version of protobuf.
2. contributors can discover which version to use by looking at the CI config.
3. if the proto files were ever changed without generating the code, we would catch the mistake immediately in the PR which made the change.
4. nothing unintentional was hidden in the generated code files (because generated code is rarely reviewed in-depth).

The CI check is currently failing. I guess because `protobuf` `v3.12.3` is not the version that was used to generate these files? Do we have a specific version to use, or should I update these files using `v3.12.3`.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -85,6 +85,36 @@ jobs:
...
+            sudo unzip -d /usr/local protoc-*.zip
+            sudo chmod +x /usr/local/bin/protoc
+            rm protoc-*.zip
+      - run:
+          name: Install gogo/protobuf
+          command: |
+            gogo_version=$(go list -m github.com/gogo/protobuf | awk '{print $2}')
+            mkdir -p .gotools; cd .gotools; go mod init consul-tools
+            go get -v github.com/hashicorp/protoc-gen-go-binary
+            go get -v github.com/gogo/protobuf/protoc-gen-gofast@${gogo_version}
```
Comment (dnephin) [inline .circleci/config.yml:108]: There is a `make tools` target, but that installs a bunch of things that aren't necessary here and adds more time to CI. Writing this kind of bash script in a makefile is not ideal. I could see it being a separate bash script, but we don't really have a pattern for that right now, other than the ones in `.circleci`.

File: build-support/scripts/proto-gen.sh [hunk 2]
```diff
@@ -116,7 +116,7 @@ function main {
       return 1
    fi
 
-   BUILD_TAGS=$(sed -e '/^[:space:]*$/,$d' < "${proto_path}" | grep '// +build')
+   BUILD_TAGS=$(sed -e '/^[[:space:]]*$/,$d' < "${proto_path}" | grep '// +build')
```
Comment (dnephin) [inline build-support/scripts/proto-gen.sh:119]: The `[:space:]` pattern does not work on linux. Could someone on macOS check this? If it doesn't work with this change we can use something like `[ \t]` I think.

Comment (mkeeler) [reply build-support/scripts/proto-gen.sh:119]: Seems to still work for me on macOS.

File: .circleci/config.yml [hunk 3]
```diff
@@ -85,6 +85,36 @@ jobs:
...
+
+      - run:
+          command: make --always-make proto
```
Comment (mkeeler) [inline .circleci/config.yml:111]: Somehow I never knew this flag existed. There are `proto-rebuild` and `proto-delete` targets in the makefile you can get rid of too.

Comment (dnephin) [reply .circleci/config.yml:111]: Oh, nice! I will remove those targets!

**Comment by dnephin** [comment]
I've also added another commit which re-generated the protobuf files using protobuf 1.12.3. Let's see if CI is happy.

**Pull request merged**